discovery/gossiper.go

```diff
@@ -2371,11 +2371,12 @@ func (d *AuthenticatedGossiper) handleChanAnnouncement(nMsg *networkMsg,
 			// If while processing this rejected edge, we realized
 			// there's a set of announcements we could extract,
 			// then we'll return those directly.
-			if len(anns) != 0 {
-				nMsg.err <- nil
-				return anns, true
-			}
+			//
+			// NOTE: since this is an ErrIgnored, we can return
+			// true here to signal "allow" to its dependants.
+			nMsg.err <- nil
```

Comment:
I could be totally wrong, but don't see how the steps you laid out give the issue?

Flow:
- `handleChanAnnouncement` in `gossiper.go`, goroutine1
  - `AddEdge` in `gossiper.go`, goroutine1
  - `InitJobDependencies` in `router.go`, goroutine2
  - `processUpdate` in `router.go`, goroutine3
    - If this returns `ErrIgnored` due to the edge already existing or being a zombie, then `SignalDependants(, allow=true)` is called. The validation barrier dependency map is cleaned up here.
  - `processRejectedEdge` in `gossiper.go`, goroutine1. If the announcement already has `AuthProof`, this will `return nil, nil`. This means `handleChanAnnouncement` will `return nil, false`
- `SignalDependants(, allow=false)` is called. However, the dependencies have already been cleaned up.

I think the issue might occur in a slightly different scenario, but can't see how it applies in the specific scenario laid out in the commit message.

From commit message:
```
Bob's ChannelAnnouncement is processed by router, and returned
  `ErrIgnored` as the edge info is already in db, and reported back to
  gossiper, the validation barrier will signal deny to all the
  ChannelUpdates jobs.
```
The router cannot fail parent validation here as it's a `ChannelEdgeInfo`, so the only way `ErrIgnored` is returned from `processUpdate` is if the edge:
- already exists, or
- is a zombie
In either case, the router code will call `SignalDependants(, allow=true)` in which case the dependencies are cleaned up and gossiper calling `SignalDependants(, allow=false)` will be a no-op

Reply:
Mostly correct, except this line,

> If this returns ErrIgnored due to the edge already existing or being a zombie, then SignalDependants(, allow=true) is called. The validation barrier dependency map is cleaned up here.

This is not necessarily true. When `ErrIgnored` is returned, you cannot be sure which one is called first,
- SignalDependants(, allow=true) in router, or,
- SignalDependants(, allow=false) in gossip

This is because, in goroutine3, we have,
1. `processUpdate`, which adds the edge info and may return an error.
2. the err is sent back to the `update.err`, which can trigger `SignalDependants(, allow=false)` in gossip.
3. SignalDependants(, allow=true) once the above error is sent.

Here are the [trimmed logs](https://gist.github.com/yyforyongyu/8292d739ee807bcc09abf226f4fad4e9), hopefully can make it more clear.

Reply:
Oh I see